### ISSUE TITLE
New Lint: `drop_non_send`

### ISSUE THREAD

**Issue by Qwaz** [rust-lang/rust-clippy#7703]
### What it does

Warns about a field in a `Send` struct that is neither `Send` nor `Copy`.

Sending the struct to another thread and drops it there will also drop the field in the new thread. This effectively changes the ownership of the field type and breaks the non-`Send` requirement which is unsound.

This is one of the most common unsound bug patterns we observed during [Rudra](https://github.com/sslab-gatech/Rudra) project.

Bug examples:

* [disrustor](https://github.com/sklose/disrustor/blob/966b949dae93d7c56f796d5c5f49be8e5a73a9c6/src/ringbuffer.rs#L4-L8) / RUSTSEC-2020-0150
* [noise_search](https://github.com/pipedown/noise/blob/5a582a7d0af1d7038646bfacaced40c3c19d4421/src/index.rs#L395-L398) / RUSTSEC-2020-0141
* [async-coap](https://github.com/google/rust-async-coap/blob/6a7b592a23de0c9d86ca399bf40ecfbf0bff6e62/async-coap/src/arc_guard.rs#L128-L132) / RUSTSEC-2020-0124
* [rusb](https://github.com/a1ien/rusb/blob/12ee91dd081428c610d3ccb5bba422198beee084/src/device_handle.rs#L107-L112) / RUSTSEC-2020-0098

### Categories (optional)

- Kind: Nursery (for now)

### Known Problems

* Raw pointers in data structures could cause false positives, such as `Vec<*const T>`. They can be partially mitigated by checking if any pointer type is used as a generic type parameter.
* A field that implements `Copy` but not `Send` can actually be a problem, e.g., [RUSTSEC-2020-0136](https://rustsec.org/advisories/RUSTSEC-2020-0136.html). However, they would be better handled in a separate lint because this lint focuses on unsoundness in drop and removing the `Copy` bound would cause too much false positives.

### Example

```rust
// Example from google/async-coap, RUSTSEC-2020-0124
#[derive(Debug, Clone)]
pub struct ArcGuard<RC, T> {
    inner: T,
    head: Arc<RC>, // This drops Arc<RC> (and in turn RC) which might not be `Send`
}

unsafe impl<RC, T: Send> Send for ArcGuard<RC, T> {} // There is no `RC: Send` bound here
```

@rustbot claim
